README.md
# Docker for Symfony (PHP-FPM - NGINX - MySQL)

Docker symfony gives you everything you need for developing Symfony application. This complete stack run with docker and docker-compose.

# Installation

## Install the latest Docker CE version
If you are running on Linux:
```
curl -fsSL https://get.docker.com -o get-docker.sh &&
sh get-docker.sh
```

If you would like to use Docker as a non-root user, you should now consider
adding your user to the "docker" group with something like:

```
sudo usermod -aG docker ${USER}
```

## Install the latest Docker Compose

If you are running on Linux:
```
sudo curl -L "https://github.com/docker/compose/releases/download/v2.2.2/docker-compose-$(uname -s)-$(uname -m)" -o /usr/local/bin/docker-compose && sudo chmod +x /usr/local/bin/docker-compose
```

## Install PWGen
### Debian / Ubuntu
```
sudo apt-get install pwgen
``` 
### CentOS
```
wget http://download-ib01.fedoraproject.org/pub/epel/7/x86_64/Packages/p/pwgen-2.08-1.el7.x86_64.rpm && rpm -ivh pwgen-2.08-1.el7.x86_64.rpm
```

# Configuration
1. Create a `.env` from the `.env.dist` file. Adapt it according to your symfony application

    ```bash
    cp .env.dist .env
    ```
   
2. Build / run containers
    ```
    ./d4d start
    ```  

3. Update your system host file (add symfony.local)
   ```bash
   # UNIX only: get containers IP address and update host (replace IP according to your configuration)
   $ sudo echo $(docker network inspect bridge | grep Gateway | grep -o -E '[0-9\.]+') "symfony.local" >> /etc/hosts
   ```
   
   **Note:** For **OS X**, please take a look [here](https://docs.docker.com/docker-for-mac/networking/).
   
4. Prepare Symfony app
    1. Get your logins to access MySQL server
    `./d4d passwd show`
    
    2. Update
  
        a) SF2, SF3: app/config/parameters.yml
          
        ```
        # ./project/app/config/parameters.yml
        parameters:
            database_host:     mysql
            database_port:     ~
            database_name:     db_name
            database_user:     db_user
            database_password: db_password (random password)
        ```
    
        b) SF4, SF5, SF6: .env
        ```
        DATABASE_URL=mysql://db_user:db_password@mysql:3306/db_name
        MAILER_URL=smtp://mailhog:1025
        ```
    3. Composer install & create database
        ```bash
        $ docker-compose exec php bash
        $ composer create-project symfony/website-skeleton my-project
            
        # Symfony 2
        $ sf doctrine:database:create
        $ sf doctrine:schema:update --force
        # Only if you have `doctrine/doctrine-fixtures-bundle` installed
        $ sf doctrine:fixtures:load --no-interaction
            
        # Symfony 3
        $ sf3 doctrine:database:create
        $ sf3 doctrine:schema:update --force
        # Only if you have `doctrine/doctrine-fixtures-bundle` installed
        $ sf3 doctrine:fixtures:load --no-interaction
    
        # Symfony 4
        $ sf4 doctrine:database:create
        $ sf4 doctrine:schema:update --force
        # Only if you have `doctrine/doctrine-fixtures-bundle` installed
        $ sf4 doctrine:fixtures:load --no-interaction
        
        # Symfony 5
        $ sf5 doctrine:database:create
        $ sf5 doctrine:schema:update --force
        # Only if you have `doctrine/doctrine-fixtures-bundle` installed
        $ sf5 doctrine:fixtures:load --no-interaction
       
        # Symfony 6
        $ sf6 doctrine:database:create
        $ sf6 doctrine:schema:update --force
        # Only if you have `doctrine/doctrine-fixtures-bundle` installed
        $ sf6 doctrine:fixtures:load --no-interaction
       ```
5. Enjoy :-)
    
    ## Usage
    
    Just run `./d4d start`, then:
    
    * Symfony app: visit [symfony.local](http://symfony.local)  
    * Symfony dev mode: visit [symfony.local/app_dev.php](http://symfony.local/app_dev.php)  
    * Logs (files location): logs/nginx and logs/symfony
    
    ## How it works?
    
    Have a look at the `docker-compose.yml` file, here are the `docker-compose` built images:
 
    * `nginx`: Nginx is one of the most popular web servers in the world and responsible for hosting some of the largest and highest-traffic sites on the internet. It is more resource-friendly than Apache in most cases and can be used as a web server or reverse proxy.
    * `php`: PHP is a popular general-purpose scripting language that is especially suited to web development.
    * `mysql:` MySQL is the most popular relational database management system.
    * `phpmyadmin`: phpMyAdmin was created so that users can interact with MySQL / MariaDB through a web interface.
    * `mailhog`: MailHog is an email testing tool for developers.
    * `redis`: Redis is an open source (BSD licensed), in-memory data structure store, used as a database, cache and message broker. It supports data structures such as strings, hashes, lists, sets, sorted sets with range queries, bitmaps, hyperloglogs, geospatial indexes with radius queries and streams. Redis has built-in replication, Lua scripting, LRU eviction, transactions and different levels of on-disk persistence, and provides high availability via Redis Sentinel and automatic partitioning with Redis Cluster. 
     
    This results in the following running containers:
    
    ```bash
    $ docker-compose ps
               Name                          Command               State                       Ports                     
    ---------------------------------------------------------------------------------------------------------------------
    docker-symfony_mailhog_1      MailHog                          Up      0.0.0.0:1025->1025/tcp, 0.0.0.0:8025->8025/tcp
    docker-symfony_mysql_1        docker-entrypoint.sh --cha ...   Up      0.0.0.0:3306->3306/tcp                        
    docker-symfony_nginx_1        nginx -g daemon off;             Up      0.0.0.0:80->80/tcp                            
    docker-symfony_php_1          docker-php-entrypoint php-fpm    Up      9000/tcp                                      
    docker-symfony_phpmyadmin_1   /docker-entrypoint.sh apac ...   Up      0.0.0.0:8080->80/tcp                          
    docker-symfony_redis_1        docker-entrypoint.sh redis ...   Up      6379/tcp          
    ```


# JetBrains support us! 

<p align="center">
  <a href="https://www.jetbrains.com" target="_blank">
      <img src="https://account.jetbrains.com/static/images/jetbrains-logo-inv.svg" width="120" title="hover text">
  </a>
</p>

# Hostinger.lt support us!

<p align="center">
  <a href="https://www.hostinger.lt/vasilij" target="_blank">
      <img src="https://www.prado.lt/wp-content/uploads/banners/hostinger.png" width="120" title="hover text">
  </a>
</p>

CHANGELOG.md
CHANGELOG for "Docker for Symfony"
======================

* 1.0.6 (2020-01-26)
    * FEATURE   [PHP] Added ability to use imagecreatefromjpeg
    * FEATURE   [PHP] Added ability to change PHP version using command ./docker set php_version 5.6 / 7.0 / 7.2 / 7.3 / 7.4
    * FEATURE   [PHP] Added ability enable / disable xdebug using command ./docker set xdebug yes / no
    * FEATURE   [PHP] Added ability enable / disable ionCube using command ./docker set ioncube yes / no
    * BUGFIX    [PHP] Returned mcrypt extension into d4d.lt installation script

* 1.0.7 (2020-02-16)
    * BUGFIX    [PHP]   Remove xdebug from pecl installation if disabled
    * FEATURE   [MYSQL] Import MySQL database copy without using a password form the console 
    * FEATURE   [PHP]   Added ability to change max_execution_time in file .env
    * FEATURE   [NGINX] Added ability to put rewrite rules into file config/nginx/rewrite/project.conf
    * FEATURE   [NGINX] Added ability select SF vs SF + PWA configuration
    * FEATURE   [PMA]   Added phpMyAdmin auto login functionality
    
* 1.0.8 (2020-04-19)
    * FEATURE   [D4D]   Added ability enable / disable redis in docker-compose.yml file
    * FEATURE   [D4D]   Added ability enable / disable rabbitmq in docker-compose.yml file
    
* 1.0.9 (2020-06-12)
    * FEATURE   [D4D]   Added option to select project type Symfony (SF), Symfony + PWA (SF_PWA), WordPress (WP)

* 1.0.10 (2020-07-22)
    * BUGFIX    [PHP]           Fix  missing redis module
    * FEATURE   [SUPERVISOR]    Added ability to switch on / off supervisor in the file .env
    * FEATURE   [ES]            Added ability to use elastic search.
    * FEATURE   [NGROK]         Added ability to use remote access.

* 1.0.11 (2020-08-26)
    * FEATURE   [D4D]           Layer optimization.
    * FEATURE   [D4D]           Added ability select MySQL vs MariaDB.
    * FEATURE   [MONGODB]       Added ability to use MongoDB.
    
* 1.0.12 (2020-11-23)
    * FEATURE   [MKCERT]        Added ability to have local developer SSL certificate.
    * FEATURE   [NGROK]         Added ability to have ngrok auth token in the file .env
    * FEATURE   [NGINX]         Updated Debian OS + nginx version.
    * FEATURE   [PHP]           The PHP 7.4 is set as default version.
    * FEATURE   [COMPOSER]      Added ability to select composer version 1.x vs 2.x
    * FEATURE   [WKHTMLTOPDF]   Added ability to select version 0.12.3, 0.12.4, 0.12.5 or 0.12.6

* 1.0.13 (2020-12-25)
  * FEATURE     [PHP]           PHP 8 is available now!
  * FEATURE     [PHP]           Added imagick changes required to support PHP 8.0
  * FEATURE     [D4D]           Added phpMyAdmin auto login functionality as root user.
  * FEATURE     [D4D]           Added ability to store .env.git without credentials (.env.secret.dist)
  * FEATURE     [D4D]           Added ability to run multiple dockers on the same time (network name external-d4d and needed to have nginx reverse proxy docker container)
  * FEATURE     [D4D]           Added ability to see running dockers IP address, version
  * FEATURE     [D4D]           Added ability to update main docker images from repository
  
* 1.0.14 (2021-02-19)
  * FEATURE     [D4D]           Added ability to see the main docker information (docker, docker-machine, docker-compose versions), main software versions, ip addresses every container
  * FEATURE     [PHP]           PHP 8 - RabbitMQ amqp extension support
  * FEATURE     [NGINX]         nginx version update
  * FEATURE     [COMPOSER]      Default composer version 2.x
  * FEATURE     [PHP]           Changed default setting for "post_max_size" and "upload_max_filesize"

* 1.0.15 (2021-04-21)
  * FEATURE     [PHP]           PHP - enabled XSL extension.
  * FEATURE     [PHP]           Added ability contributing to Symfony.
  * FEATURE     [D4D]           Added ability run default container on ./d4d start.
  * FEATURE     [NGINX]         NGINX version update.
  * FEATURE     [PHP]           Blackfire Agent v1 to v2 upgrade.

* 1.0.16 (2021-07-18)
  * FEATURE     [PHP]           Added PHP-CS-Fixer implementation.
  * FEATURE     [D4D]           Updated docker-compose version.
  * BUGFIX      [D4D]           NGINX + PWA network issue bug fix.
  * BUGFIX      [PHP]           PHP configuration bugfix.
  * FEATURE     [PHP]           Added Local PHP Security Checker implementation.
  * FEATURE     [NGINX]         NGINX version update.
  * FEATURE     [COMPOSER]      Composer version update.
  * FEATURE     [MySQL]         Default MySQL version as MariaDB.
  * FEATURE     [D4D]           Added webpack-analyzer.
  * FEATURE     [NGINX]         NGINX version update.
  * FEATURE     [D4D]           Added Symfony CLI implementation.

* 1.0.17 (2021-10-05)
  * FEATURE     [RabbitMQ]      Updated RabbitMQ version.
  * FEATURE     [D4D]           Updated Symfony CLI version.
  * FEATURE     [NGINX]         Updated Nginx version.
  * BUGFIX      [D4D]           MySQL 8.0 version not shown issue fix.
  * FEATURE     [D4D]           Docker code optimization.
  * FEATURE     [MariaDB]       Updated MariaDB version.
  * FEATURE     [D4D]           Updated composer version.
  * FEATURE     [PHP]           Added eslint implementation.
  * FEATURE     [PHP]           Added PHPStan integration.
  * FEATURE     [NGINX]         Nginx layer optimization.
  * FEATURE     [D4D]           mysql_dump.sh allow non-root user.

* 1.0.18 (2021-12-22)
  * FEATURE     [COMPOSER]      Updated composer version.
  * FEATURE     [D4D]           Added Deployer.org implementation.
  * FEATURE     [D4D]           Removed docker-machine from installation script.
  * BUGFIX      [D4D]           Fix container`s IP addresses.
  * FEATURE     [D4D]           Added docker container backup implementation.
  * FEATURE     [RabbitMQ]      Updated RabbitMQ version.
  * FEATURE     [PHP]           Updated Symfony CLI version.
  * FEATURE     [D4D]           Updated D4D installation script.

docker.sh
#!/bin/bash

DOCKER_SCRIPT_VER=1.0.0

WORK_DIR=`pwd`

cd ${WORK_DIR}

#Added for security
chmod 700 ${WORK_DIR}/docker.sh >/dev/null 2>&1
chmod 700 ${WORK_DIR} >/dev/null 2>&1

ENV=${WORK_DIR}/.env
ENV_DIST=${WORK_DIR}/.env.dist

if [ ! -e ${ENV} ]; then
  cp "${ENV_DIST}" "${ENV}"
fi

# Emulate ${!variable}
eval_var() {
    var=${1}
        if [ -z ${var} ]; then
            echo ""
        else
            eval newval="\$${var}"
            echo $newval
        fi
}

HIDE_CHANGES=0

getOpt() {
	#$1 is option name
	#$2 is default value

  source ${WORK_DIR}/.env

	GET_OPTION="$(eval_var $1)"
	if [ "${GET_OPTION}" = "" ]; then
		echo "$1=$2" >> ${ENV}
		GET_OPTION="${2}"
		eval `echo "${1}=${2}"`
	fi

	echo ${GET_OPTION}
}

##################################################
# Variables for bolded text
boldon="`tput -Txterm bold`"
boldoff="`tput -Txterm sgr0`"
##################################################

setOpt() {
	#$1 is option name
	#$2 is value
	READ_OPTION_NAME="$1"
	READ_OPTION_VALUE="`echo \"$2\" | perl -p0 -e 's|@|\\\\@|g'`"

	VAR=`echo ${READ_OPTION_NAME} | tr "[a-z]" "[A-Z]"`
	if [ -z "$(eval_var ${VAR}_DEF)" ]; then
		echo "${READ_OPTION_NAME} is not a valid option."
		EXIT_CODE=50
		return
	fi

	VALID="no"
	for i in $(eval_var ${VAR}_SET); do
		if [ "${i}" = "${READ_OPTION_VALUE}" ] || [ "${i}" = "userinput" ]; then
			VALID="yes"
			break
		fi
	done

	if [ "${VALID}" = "no" ]; then
		echo "${READ_OPTION_VALUE} is not a valid setting for ${READ_OPTION_NAME} option."
		EXIT_CODE=51
		return
	fi
	OPT_VALUE="`grep -m1 "^${READ_OPTION_NAME}=" ${ENV} | cut -d= -f2 | perl -p0 -e 's|@|\\\\@|g'`"
	perl -pi -e "s#${READ_OPTION_NAME}=${OPT_VALUE}#${READ_OPTION_NAME}=${READ_OPTION_VALUE}#" ${ENV}
	if [ "${HIDE_CHANGES}" = "0" ]; then
		echo "Changed ${boldon}${READ_OPTION_NAME}${boldoff} option from ${boldon}${OPT_VALUE}${boldoff} to ${boldon}${READ_OPTION_VALUE}${boldoff}" | perl -p0 -e 's|\\\@|\@|g'
	fi
}

##################################################

PHP_VERSION_SET="5.6 7.0 7.1 7.2 7.3"

PHP_VERSION_DEF="7.3"

# PHP
PHP_VERSION_OPT=`getOpt PHP_VERSION ${PHP_VERSION_DEF}`

##################################################

# Help sections
HELP_SECTIONS="DOCKER_RUN DOCKER_MYSQL DOCKER_PASSWD DOCKER_D4D"
DOCKER_RUN_SET="start stop exec logs"
DOCKER_RUN_DESC="Docker for Symfony (PHP-FPM - NGINX - MySQL)"
DOCKER_RUN_REQADD="start stop exec logs"
START_ADDIT=""
STOP_ADDIT=""
EXEC_ADDIT=""
LOGS_ADDIT=""

DOCKER_MYSQL_SET="mysql_dump mysql_restore"
DOCKER_MYSQL_DESC="Backup and Restore a MySQL Database"
DOCKER_MYSQL_REQADD="mysql_dump mysql_restore"
DOCKER_MYSQL_DUMP_ADDIT=""
DOCKER_MYSQL_RESTORE_ADDIT=""

DOCKER_PASSWD_SET="passwd"
DOCKER_PASSWD_DESC="Show password sensitive information"
DOCKER_PASSWD_REQADD="passwd"
PASSWD_ADDIT="show"

DOCKER_D4D_SET="opt_help set"
DOCKER_D4D_DESC="Docker related options/functions"
DOCKER_D4D_REQADD="opt_help set"
OPT_HELP_ADDIT="(full)"
SET_ADDIT="option_name value"

##################################################

# ALL SETTINGS
# SECTIONS OF OPTIONS
ALL_SECTIONS="PHP_SETTINGS"

PHP_SETTINGS="PHP_VERSION"

showVersion() {
    echo "${DOCKER_SCRIPT_VER} (rev: 0001)"
}

generateHelp() {
    for section in ${HELP_SECTIONS}; do
        DESC=${section}_DESC
        echo " +-----------------------------------------------------------+"
        printf " | %-55s %-2s|\n" "$(eval_var ${DESC}):"

        BUILDSET="${section}_SET"
        BUILD_ADD_TO_OUT="${section}_REQADD"
        for setting in $(eval_var ${BUILDSET}); do
            ADDIT=""
            for i in $(eval_var ${BUILD_ADD_TO_OUT}); do
                ADDIT_VAR=`echo "${i}_ADDIT" | tr "[a-z]" "[A-Z]"`
                if [ "$i" = "${setting}" ]; then
                    ADDIT="$(eval_var ${ADDIT_VAR})"
                fi
                done
                printf " | %-55s %-2s|\n" "   $0 ${setting} ${ADDIT}"
            done
    done
}

doChecks() {
    # Prerequisites
    pwgen > /dev/null 2>&1 || { echo >&2 "pwgen not found."; exit 1; }
    docker --version > /dev/null 2>&1 || { echo >&2 "Docker not found. Please install it via https://docs.docker.com/install/"; exit 1; }
    docker-machine --version > /dev/null 2>&1 || { echo >&2 "Docker machine not found. https://docs.docker.com/machine/install-machine/"; exit 1; }
    docker-compose --version > /dev/null 2>&1 || { echo >&2 "Docker compose not found. Please install it via https://docs.docker.com/compose/install/"; exit 1; }

    if [ ! -f "${ENV}" ]; then
        if [ -f "${ENV_DIST}" ]; then
            cp "{$ENV_DIST}" "${ENV}"
        else
            echo >&2 "The ${ENV} file does not exist. Project setup will not work"
            exit 1
        fi
    fi

    # Assign user id and group id into variables
    TMP_USER_ID=`id -u`
    TMP_GROUP_ID=`id -g`

    # Always validate user id and group id before start using .env file
    sed -i 's#USER_ID=.*#'"USER_ID=${TMP_USER_ID}"'#g' ${WORK_DIR}/.env
    sed -i 's#GROUP_ID=.*#'"GROUP_ID=${TMP_USER_ID}"'#g' ${WORK_DIR}/.env
    sed -i 's#MYSQL_ROOT_PASSWORD=root#'"MYSQL_ROOT_PASSWORD=`pwgen -s 20 1`"'#g' ${WORK_DIR}/.env
    sed -i 's#MYSQL_PASSWORD=db_password#'"MYSQL_PASSWORD=`pwgen -s 20 1`"'#g' ${WORK_DIR}/.env

    # Load .env file into the current shell script
    source ${WORK_DIR}/.env

    # Ensure all folders exists
    mkdir -p ${PROJECT_PATH}
    mkdir -p ${PROJECT_PATH}/${SYMFONY_LOG_PATH}
    mkdir -p ${SYMFONY_LOG_PATH}
    mkdir -p ${COMPOSER_PATH}
    mkdir -p ${COMPOSER_PATH}/cache
    mkdir -p ${SSH_KEY_PATH}
    mkdir -p ${MYSQL_DUMP_PATH}

    # Create an SSH private and public keys if we do not have it
    if [ ! -f "${SSH_KEY_PATH}/id_rsa" ]; then
        ssh-keygen -b 4096 -t rsa -f ${SSH_KEY_PATH}/id_rsa -q -P ""
    fi

    # Create a file if it does not exist
    if [ ! -f "${SSH_KEY_PATH}/known_hosts" ]
    then
        touch ${SSH_KEY_PATH}/known_hosts
    fi

    # Ensure all folders exists
    mkdir -p ${NGINX_LOG_PATH}
    mkdir -p ${MYSQL_DATA_PATH}
    mkdir -p ${USER_CONFIG_PATH}

    # Create a file if it does not exist
    touch ${USER_CONFIG_PATH}/.bash_history

    if [ ! -f "${USER_CONFIG_PATH}/.my.cnf" ]; then
        printf "[client]\nuser=${MYSQL_USER}\npassword=${MYSQL_PASSWORD}\n" >> ${USER_CONFIG_PATH}/.my.cnf
    fi

    #
#    ./config/php/build.sh

    docker-compose build

    # Clears the screen.
    clear
}

start() {
    # Start server
    echo "Starting docker containers..."
    docker-compose up -d

    # Documentation for end user
    echo ""
    echo "The following information has been set:"
    echo ""
    echo "Server IP: 127.0.0.1"
    echo "Server Hostname: ${PROJECT_DOMAIN}"
    echo ""
    echo "To login now, follow this link:"
    echo ""
    echo "Project URL: http://${PROJECT_DOMAIN}"
    echo "phpMyAdmin: http://${PROJECT_DOMAIN}:8080"
    echo "MailHog: http://${PROJECT_DOMAIN}:8025"
    echo ""
    echo "Thank you for using Docker for Symfony. Should you have any questions, don't hesitate to contact us at support@prado.lt"
}

stop() {
    docker-compose down
}

dockerComposeExec() {
    # $1 command
    docker-compose exec $1 bash
}

doMysqlDump() {
    # Load .env file into the current shell script
    source ${WORK_DIR}/.env

    docker-compose exec mysql bash /tmp/db/mysql_dump.sh
}

doPasswd() {
    # $1 command
    COMMAND=$1

    if [ "${COMMAND}" = "show" ]; then
        showPasswords
    fi
}

allSettings() {
	for section in $ALL_SECTIONS; do
		DESC=${section}_DESC
		echo "------------------------------------------"
		echo "$(eval_var ${DESC})"
		echo "------------------------------------------"
		for setting in $(eval_var ${section}); do
			SETTING_NAME=`echo $setting | tr "[A-Z]" "[a-z]"`
			POSSIBLE_VALUES_VAR=${setting}_SET
			POSSIBLE_VALUES="`echo $(eval_var ${POSSIBLE_VALUES_VAR}) | awk -v OFS=", " '$1=$1'`"
			DEFAULT_VALUE=${setting}_DEF
			CURRENT_VALUE=${setting}_OPT
			echo -n "${SETTING_NAME}: ${POSSIBLE_VALUES}. Current value: $(eval_var ${CURRENT_VALUE}). Default value: $(eval_var ${DEFAULT_VALUE})."
			if [ "$1" = "full" ]; then
				DESCRIPTION="${setting}_DESC"
				echo " Description: $(eval_var ${DESCRIPTION})"
			else
				echo ""
			fi
		done
		echo ""
	done
}

mainHeader() {
    printf " +%-55s+\n" "-----------------------------------------------------------"
    printf " | %-55s %-2s|\n" "Docker for Symfony (PHP-FPM - NGINX - MySQL)"
    printf " | %-55s %-2s|\n" "Written by Vasilij Dusko"
    printf " | %-55s %-2s|\n" "Version: $(showVersion)"
    printf " +%-55s+\n" "-----------------------------------------------------------"
}

showPasswords() {
    source ${WORK_DIR}/.env

    mainHeader
    printf " +%-55s+\n" "-----------------------------------------------------------"
    printf " | %-55s %-2s|\n" "The following information has been set:"
    printf " +%-55s+\n" "-----------------------------------------------------------"
    printf " | %-55s %-2s|\n" "Server IP: 127.0.0.1"
    printf " | %-55s %-2s|\n" "Server Hostname: ${PROJECT_DOMAIN}"
    printf " +%-55s+\n" "-----------------------------------------------------------"
    printf " | %-55s %-2s|\n" "MySQL root username: root"
    printf " | %-55s %-2s|\n" "MySQL root password: ${MYSQL_ROOT_PASSWORD}"
    printf " +%-55s+\n" "-----------------------------------------------------------"
    printf " | %-55s %-2s|\n" "MySQL database name: ${MYSQL_DATABASE}"
    printf " | %-55s %-2s|\n" "MySQL username: ${MYSQL_USER}"
    printf " | %-55s %-2s|\n" "MySQL password: ${MYSQL_PASSWORD}"
    printf " +%-55s+\n" "-----------------------------------------------------------"
    printf " | %-55s %-2s|\n" "To login now, follow this link:"
    printf " +%-55s+\n" "-----------------------------------------------------------"
    printf " | %-55s %-2s|\n" "Project URL: http://${PROJECT_DOMAIN}"
    printf " | %-55s %-2s|\n" "phpMyAdmin: http://${PROJECT_DOMAIN}:8080"
    printf " | %-55s %-2s|\n" "MailHog: http://${PROJECT_DOMAIN}:8025"
    printf " +%-55s+\n" "-----------------------------------------------------------"
}

showHelp() {
    mainHeader
    generateHelp
    printf " +%-55s+\n" "-----------------------------------------------------------"
}

case "$1" in
    start) doChecks; start
        ;;
    stop) stop
        ;;
    exec) dockerComposeExec $2
        ;;
    mysql_dump) doMysqlDump
        ;;
    passwd) doPasswd $2
        ;;
    set) setOpt $2 $3
        ;;
    opt_help) allSettings $2
        ;;
    * ) showHelp
        exit 0
        ;;
esac
